Add support for optional kafka SASL

ShahimSharafudeen opened the pull request:
## Description
Currently the presto-kafka connector can be configured to use SASL authentication (https://docs.confluent.io/platform/current/security/authentication/sasl/plain/overview.html#kafka-sasl-auth-plain) those config properties were made known to the KafkaConfig and are used in the producer as well as consumer factory if being set (all optional).

As of now, we have configured the following parameters for the above requirement:

1. security.protocol
2. sasl.mechanism
3. sasl.jaas.config
4. ssl.truststore.location
5. ssl.truststore.password
6. ssl.truststore.type

As part of the code-level configuration, represent the parameter naming as below and map them to the corresponding Kafka config parameters:

1. kafka.security-protocol=SASL_SSL
2. kafka.sasl.mechanism=PLAIN
3. kafka.sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username=<USERNAME> password=<API KEY/PASSWORD>;
4. kafka.truststore.path=XXX
5. kafka.truststore.password=XXX
6. kafka.truststore.type=XXX


Taking reference from Trinodb Kafka SSL implementation: https://github.com/trinodb/trino/pull/6929

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Kafka Connector Changes 
* Add support for optional Apache Kafka SASL. 
```

aaneja reviewed (changes requested):
- Can you attribute  trinodb/trino#6929 in your commit message ? See [backport commits] from the [CONTRIBUTING](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md) doc
- Can you update the `kafka.rst` file with the new SASL properties

ShahimSharafudeen commented:
> * Can you attribute  [Add support for optional Kafka SSL trinodb/trino#6929](https://github.com/trinodb/trino/pull/6929) in your commit message ? See 'backport commits' section from the [CONTRIBUTING](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md) doc
> * Can you update the `kafka.rst` file with the new SASL properties

@aaneja - Addressed the review comments.

steveburnett reviewed (changes requested):
Thanks for the doc! Looks good. Some minor formatting suggestions, and one question for you to consider.

ShahimSharafudeen commented:
> Thanks for the doc! Looks good. Some minor formatting suggestions, and one question for you to consider.

Thanks @steveburnett for reviewing the documentation-related changes. I’ve addressed all your review comments.

ShahimSharafudeen commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -28,11 +28,17 @@ To configure the Kafka connector, create a catalog properties file
 replacing the properties as appropriate:
 
+In some cases, such as when using specialized authentication methods, it is necessary to specify
+additional Kafka client properties in order to access your Kafka cluster. To do so,
```
Done.

ShahimSharafudeen commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -28,11 +28,17 @@ To configure the Kafka connector, create a catalog properties file
 
+In some cases, such as when using specialized authentication methods, it is necessary to specify
+additional Kafka client properties in order to access your Kafka cluster. To do so,
+add the `kafka.config.resources` property to reference your Kafka config files. Note that configs
```
Done

ShahimSharafudeen commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -28,11 +28,17 @@ To configure the Kafka connector, create a catalog properties file
+In some cases, such as when using specialized authentication methods, it is necessary to specify
+additional Kafka client properties in order to access your Kafka cluster. To do so,
+add the `kafka.config.resources` property to reference your Kafka config files. Note that configs
+can be overwritten if defined explicitly in `kafka.properties`:
```
Done

ShahimSharafudeen commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -28,11 +28,17 @@ To configure the Kafka connector, create a catalog properties file
+In some cases, such as when using specialized authentication methods, it is necessary to specify
+additional Kafka client properties in order to access your Kafka cluster. To do so,
+add the `kafka.config.resources` property to reference your Kafka config files. Note that configs
+can be overwritten if defined explicitly in `kafka.properties`:
```
`Does this mean "Configuration settings in the files in kafka.config.resources can be overridden if specified explicitly in kafka.properties." ?`

Yes, I have updated the sentence to make it a bit more meaningful and clear, as shown above, and provided an example.

ShahimSharafudeen commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -137,6 +150,50 @@ show up in ``DESCRIBE <table-name>`` or ``SELECT *``.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Protocol used to communicate with brokers.
+Valid values are: SASL_PLAINTEXT, SASL_SSL.
```
Done.

ShahimSharafudeen commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -137,6 +150,50 @@ show up in ``DESCRIBE <table-name>`` or ``SELECT *``.
+``kafka.sasl.mechanism``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This specifies the authentication mechanism used by SASL.
```
Done.

ShahimSharafudeen commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -137,6 +150,50 @@ show up in ``DESCRIBE <table-name>`` or ``SELECT *``.
+
+``kafka.sasl.jaas.config``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+This config used for SASL authentication, which is a method for securing communication with brokers.
```
Done

aaneja reviewed (changes requested):
Can you add a product test that uses these new configs ? We currently have a `singlenode-kafka` test, lets make a new one called `singlenode-kafka-sasl`

steveburnett commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -28,11 +28,18 @@ To configure the Kafka connector, create a catalog properties file
 
+In some cases, such as when using specialized authentication methods, it is necessary to specify
+additional Kafka client properties to access your Kafka cluster. To do so,
+add the ``kafka.config.resources`` property to reference your Kafka config files.
```
```suggestion
add the ``kafka.config.resources`` property to reference your Kafka configuration files.
```

steveburnett commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -28,11 +28,18 @@ To configure the Kafka connector, create a catalog properties file
+additional Kafka client properties to access your Kafka cluster. To do so,
+add the ``kafka.config.resources`` property to reference your Kafka config files.
+Configuration settings in the files in ``kafka.config.resources`` can be overridden if
+specified explicitly in ``kafka.properties`` (e.g., security.protocol=SSL):
```
```suggestion
specified explicitly in ``kafka.properties`` (for example, ``security.protocol=SSL``):
```
Avoid Latin abbreviations and phrases. See the [Gitlab Documentation Style Guide Recommended Word List entry for e.g.](https://docs.gitlab.com/development/documentation/styleguide/word_list/#eg) for suggestions of alternatives.

steveburnett commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -137,6 +151,50 @@ show up in ``DESCRIBE <table-name>`` or ``SELECT *``.
+
+``kafka.sasl.jaas.config``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+This configuration used for SASL authentication, which is a method for securing communication with brokers.
```
```suggestion
The configuration used for SASL authentication, which is a method for securing communication with brokers.
```

steveburnett commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -137,6 +151,50 @@ show up in ``DESCRIBE <table-name>`` or ``SELECT *``.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+This configuration used for SASL authentication, which is a method for securing communication with brokers.
+
+This property is optional, but required when ``kafka.sasl.mechanism`` is given.
```
```suggestion
This property is optional, but is required when ``kafka.sasl.mechanism`` is given.
```

steveburnett commented on presto-docs/src/main/sphinx/connector/kafka.rst:
```diff
@@ -137,6 +151,50 @@ show up in ``DESCRIBE <table-name>`` or ``SELECT *``.
+``kafka.truststore.path``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Path of the truststore file used for connection to Kafka cluster.
```
```suggestion
Path of the truststore file used for connecting to the Kafka cluster.
```

steveburnett reviewed (changes requested):
Thanks for the quick update! I found a few more nits and suggestions, again nothing major.

ShahimSharafudeen commented:
@aaneja - I have addressed all your comments regarding the Java-related changes.

ShahimSharafudeen commented:
> Thanks for the quick update! I found a few more nits and suggestions, again nothing major.

Thanks again, @steveburnett, for your suggestions. I have addressed all your comments regarding the doc-related changes.

steveburnett reviewed (dismissed):
LGTM! (docs)

Pull updated branch, new local doc build, looks good. Thanks!

aaneja commented on presto-kafka/src/main/java/com/facebook/presto/kafka/SaslKafkaConsumerManager.java:
```diff
@@ -0,0 +1,50 @@
+{
+    private final KafkaSaslConfig saslConfig;
+    private final KafkaConsumerManager delegate;
+    @Inject
```
nit: Add a newline above

aaneja commented on presto-kafka/src/main/java/com/facebook/presto/kafka/SaslKafkaProducerFactory.java:
```diff
@@ -0,0 +1,49 @@
+    private final KafkaSaslConfig saslConfig;
+    private final KafkaProducerFactory delegate;
+
+    private static final Logger log = Logger.get(SaslKafkaProducerFactory.class);
```
Unused, please remove

aaneja commented on presto-kafka/src/main/java/com/facebook/presto/kafka/SaslKafkaProducerFactory.java:
```diff
@@ -0,0 +1,49 @@
+    private final KafkaProducerFactory delegate;
+
+    private static final Logger log = Logger.get(SaslKafkaProducerFactory.class);
+    @Inject
```
Add a newline above

ShahimSharafudeen replied:
Done.

aaneja commented on presto-kafka/src/main/java/com/facebook/presto/kafka/PlainTextKafkaConsumerManager.java:
```diff
@@ -0,0 +1,65 @@
+        final Properties properties = new Properties();
+        properties.put(BOOTSTRAP_SERVERS_CONFIG, hostAddress.toString());
+        properties.put(GROUP_ID_CONFIG, threadName);
+        properties.put(MAX_POLL_RECORDS_CONFIG, Integer.toString(maxPollRecords));
```
Is the `Integer.toString` needed ? From a quick glance at ConsumerConfig, it looks like we could use the integer directly

ShahimSharafudeen replied:
@aaneja - We can pass the value either as an int or a String, but to avoid any issues during runtime, it's better to stick with the existing code pattern used in the initial implementation of the KafkaConsumerManager class, which passes it as a String. Let me know if you’d prefer using an Integer instead — happy to update it.

aaneja commented:
Please create an issue to implement the `singlenode-kafka-sasl` product test. This can be punted to later

ShahimSharafudeen commented:
> Please create an issue to implement the `singlenode-kafka-sasl` product test. This can be punted to later

Git issue : https://github.com/prestodb/presto/issues/24993

steveburnett reviewed (approved):
LGTM! (docs) 

Pull updated branch, new local doc build, looks good.

ZacBlanco reviewed (commented):
Minor things, overall this is good

ShahimSharafudeen commented on presto-kafka/src/test/java/com/facebook/presto/kafka/TestKafkaConnectorConfig.java:
```diff
@@ -32,12 +36,20 @@ public void testDefaults()
     {
+        String secret = "confluent";
+        Path tempFile1Path = Files.createTempFile("tempFile1", ".txt");
+        Path tempFile2Path = Files.createTempFile("tempFile2", ".txt");
```
The `kafka.config.resources` parameter expects a file path as a string value. In this test case, passing either a dummy file name or a real file path works fine. However, since the file contents are not actually used, a dummy file name is sufficient. 
Based on your suggestion and best practices, I’ve removed the unnecessary file creation steps.

ShahimSharafudeen commented on presto-kafka/src/main/java/com/facebook/presto/kafka/security/KafkaKeystoreTruststoreType.java:
```diff
@@ -0,0 +1,28 @@
+    JKS,
+    PKCS12;
+
+    public static Optional<KafkaKeystoreTruststoreType> fromString(String value)
```
Updated the fromString method to properly handle invalid input by catching exceptions and returning Optional.empty(), based on your second suggestion.

ZacBlanco replied:
Sorry for the confusion, but I am checking out this PR again. This method isn't called anywhere from what I can tell. I think we can remove it?

ShahimSharafudeen commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConnectorModule.java:
```diff
@@ -96,4 +98,21 @@ public void bindTopicSchemaProviderModule(String name, Module module, Function<K
+    {
+        install(installModuleIf(
+                KafkaSecurityConfig.class,
+                config -> (config.getSecurityProtocol().isPresent() && config.getSecurityProtocol().get().equals(SecurityProtocol.SASL_SSL)),
```
Done.

ZacBlanco commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConnectorModule.java:
```diff
@@ -96,4 +99,21 @@ public void bindTopicSchemaProviderModule(String name, Module module, Function<K
+    {
+        install(installModuleIf(
+                KafkaSecurityConfig.class,
+                config -> (config.getSecurityProtocol().isPresent() && config.getSecurityProtocol().get().equals(SASL_SSL)),
```
```suggestion
                config -> config.getSecurityProtocol().isPresent() && config.getSecurityProtocol().get().equals(SASL_SSL),
```

ShahimSharafudeen replied:
Done

ZacBlanco commented on presto-kafka/src/main/java/com/facebook/presto/kafka/KafkaConnectorModule.java:
```diff
@@ -96,4 +99,21 @@ public void bindTopicSchemaProviderModule(String name, Module module, Function<K
+    {
+        install(installModuleIf(
+                KafkaSecurityConfig.class,
+                config -> (config.getSecurityProtocol().isPresent() && config.getSecurityProtocol().get().equals(SASL_PLAINTEXT)
+                        || !config.getSecurityProtocol().isPresent()),
```
```suggestion
                config -> config.getSecurityProtocol().isPresent() && config.getSecurityProtocol().get().equals(SASL_PLAINTEXT)
                        || !config.getSecurityProtocol().isPresent(),
```

ShahimSharafudeen replied:
Done.

ZacBlanco reviewed (changes requested):
Can you also rebase this PR on the latest master branch? I noticed the git log doesn't include anything from April

ShahimSharafudeen commented:
> Can you also rebase this PR on the latest master branch? I noticed the git log doesn't include anything from April

Rebased the PR with the latest commits from the master branch.

steveburnett commented:
Thanks for the release note! Minor nit suggestions: 

```
== RELEASE NOTES ==

Kafka Connector Changes
* Add support for optional Apache Kafka SASL. 
```

ShahimSharafudeen commented:
> Thanks for the release note! Minor nit suggestions: 
> 
> ```
> == RELEASE NOTES ==
> 
> Kafka Connector Changes
> * Add support for optional Apache Kafka SASL. 
> ```
Done
